pkg/kubelet/prober/runner.go

```diff
@@ -0,0 +1,215 @@
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+	req, err := httpprobe.NewRequestForHTTPGetAction(probeSpec.HTTPGet, &container, status.PodIP, "probe")
+	if err != nil {
+		return err
+	}
+
+	// persist
+	hp.timeout = timeout
+	hp.httpRequest = req
+	return err
+}
+
+func (hp *httpProbeRunner) run(_ context.Context, prober prober) (probe.Result, string, error) {
+	//return prober.(httpprobe.Prober).Probe(hp.httpRequest, hp.timeout)
+	return prober.http.Probe(hp.httpRequest, hp.timeout)
+}
+
+// grpcProbeRunner holds state and logic for running GRPC probes.
+type grpcProbeRunner struct {
+	host    string
+	service string
+	port    int
+	timeout time.Duration
+}
+
+// newGRPCProbeRunner returns grpcProbeRunner which implements probeRunner.
+func newGRPCProbeRunner() *grpcProbeRunner {
+	return &grpcProbeRunner{}
+}
+
+func (gp *grpcProbeRunner) init(probeSpec v1.Probe, container v1.Container, status v1.PodStatus) error {
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	host := status.PodIP
+	service := ""
+	if probeSpec.GRPC.Service != nil {
+		service = *probeSpec.GRPC.Service
+	}
+
+	// persist
+	gp.host = host
+	gp.service = service
+	gp.port = int(probeSpec.GRPC.Port)
+	gp.timeout = timeout
+	return nil
+}
+
+func (gp *grpcProbeRunner) run(_ context.Context, prober prober) (probe.Result, string, error) {
+	//return prober.(grpcprobe.Prober).Probe(gp.host, gp.service, gp.port, gp.timeout)
+	return prober.grpc.Probe(gp.host, gp.service, gp.port, gp.timeout)
+}
+
+// execProbeRunner holds state and logic for running Exec probes.
+type execProbeRunner struct {
+	containerID kubecontainer.ContainerID
+	command     []string
+	timeout     time.Duration
+}
+
+// newExecProbeRunner returns execProbeRunner which implements probeRunner.
+func newExecProbeRunner() *execProbeRunner {
+	return &execProbeRunner{}
+}
+
+func (ep *execProbeRunner) init(probeSpec v1.Probe, container v1.Container, status v1.PodStatus) error {
+	command := kubecontainer.ExpandContainerCommandOnlyStatic(probeSpec.Exec.Command, container.Env)
+	timeout := time.Duration(probeSpec.TimeoutSeconds) * time.Second
+
+	c, _ := podutil.GetContainerStatus(status.ContainerStatuses, container.Name)
+	containerID := kubecontainer.ParseContainerID(c.ContainerID)
+
+	// persist
+	ep.command = command
+	ep.timeout = timeout
+	ep.containerID = containerID
+	return nil
+}
+
+func (ep *execProbeRunner) run(ctx context.Context, prober prober) (probe.Result, string, error) {
+	return prober.exec.Probe(&execInContainer{run: func() ([]byte, error) {
+		return prober.runner.RunInContainer(ctx, ep.containerID, ep.command, ep.timeout)
+	}})
+}
+
+// newProbeRunner is a factory which returns concrete implementation of probeRunner
+// depending on the kind of the probe - tcp, http, grpc and exec.
+func newProbeRunner(container v1.Container, probeType probeType) (probeRunner, error) {
```

Comment:
@SergeyKanzhelev is it fine if `newProbeRunner ` doesn't return any error, just logs the error cases.
`newWorker` doesn't return any error, and I have to use `newProbeRunner ` in `newWorker`. 

I'm not sure what to do in cases of unknown kind of probes or nil probes.

Reply:
good question. Are those cases only hypothetical? If those are validated by the API validation we should have an empty prober that will emit warnings instead of executing probes. Like suggested here: https://github.com/kubernetes/kubernetes/issues/116026 - all Unknown cases needs to result in Warning events

Reply:
yes, these are hypothetical cases only. 

for the run() func of this new empty prober, should I just return the status and error, and wait for development around https://github.com/kubernetes/kubernetes/issues/116026 ?

Reply:
Yes those are hypothetical cases only. I added an `emptyProberRunner`, and will wait for development around `https://github.com/kubernetes/kubernetes/issues/116026` to finalize the warning events.

Reply:
I think your PR will change things significantly. So need to address https://github.com/kubernetes/kubernetes/issues/116026 as well. I suggest to have a new Prober type that instead of probes just reports a warning error. Create and return it from the factory if you failed to contruct a proper object

Reply:
I already pushed the `emptyProbeRunner` object, my only concern is that `func (pb *prober) probe()` handles all error and event logging (after reattempts), and emitting warnings from the `emptyProbeRunner` directly will result in duplicate code at two places. Right now I'm just returning error when `run()` is called on `emptyProbeRunner` and `func (pb *prober) probe()` can handle the logging.

Let me know your thoughts!